pinot-core/src/main/java/org/apache/pinot/core/segment/processing/partitioner/TableConfigPartitioner.java

```diff
@@ -0,0 +1,45 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.segment.processing.partitioner;
+
+import org.apache.pinot.core.data.partition.PartitionFunction;
+import org.apache.pinot.core.data.partition.PartitionFunctionFactory;
+import org.apache.pinot.spi.config.table.ColumnPartitionConfig;
+import org.apache.pinot.spi.data.readers.GenericRow;
+
+
+/**
+ * Partitioner which computes partition values based on the ColumnPartitionConfig from the table config
+ */
+public class TableConfigPartitioner implements Partitioner {
```

Comment:
What if I need to align data on time while the table is custom partitioned? (just trying to brainstorm how we will extend the current partitioner to support this)

Then, we can probably add the new partitioner that combines the value from `TableConfigPartitioner` and `TransformationPartitioner`?


e.g.  partition on memberId using murmur, need to enable segment merge so the data needs to be time aligned.

1. Use table config partitioner to get the partition id based on murmur on memberId -> Let's day `2`
2. Use time align partitioner -> Let's say `2020/12/12`

Combine 1&2 -> `2020/12/12-2` <- example of partitionId

We can do something like the above?

Reply:
this is a good point. I did not think of this. Yes we can add new partitioner in that case, which combines the partitioners as needed.

But this might be a common case, so I'm thinking if we should have that in the design itself. How about we always do partitioning in 2 steps inside the mapper. 
1) Apply any partitioning from Segment Processor Config 
2) Apply any partitioning from Table Config. 
So in your example, first TransformFunctionPartitioner gets applied and generated date partition. Then, we check if table config partitioner exists, and if yes, further apply the PartitionFunction. Then concat the 2 partitions like you suggested.

Another option is we make PartitionerConfig a list, and apply all partitioners one by one, and concat all to get final partition.